Update backProp; add new method compareOutput

yonatanmg opened the pull request:
added compareOutput to be used in backprop and/or classify that compares the last layer of activations with expected vals. thought this might make for cleaner code than having similar functions inside of backprop and classify
the stuff about checking to make sure expectedValues is not bad data isn't really necessary (yet) but it doesn't add much weight to the program so i just left it there

yonatanmg committed:
Update backProp; add new method compareOutput

added compareOutput to be used in backprop and/or classify that compares the last layer of activations with expected vals. thought this might make for cleaner code than having similar functions inside of backprop and classify
the stuff about checking to make sure expectedValues is not bad data is really necessary (yet) but it doesn't add much weight to the program so i just left it there

patelami3431 commented on Network Class/Network.cpp:
```diff
@@ -178,6 +176,60 @@ bool backProp(int index)
+	// indices at which there is the biggest activation or a 1 for the netowrk or expected vals respectively
+	int biggestI = -1; int expectedI = -1;
+
+	// biggest starts at 0 (not MIN) because any number < 0 might as well be 0 for these purposes
```
The biggest has to start with the -1 or min negative value possible because when we change the activation function we might get values from -1 to 1 (eg: tanh())

patelami3431 replied:
I am sorry, I didn't read the complete function. But yeah, that makes sense now!

kjhsdgf commented on Network Class/Network.cpp:
```diff
@@ -178,6 +176,60 @@ bool backProp(int index)
+	int lastInd = numLayers - 1;
+	int lastSize = layerSizes[lastInd];
+
+	// indices at which there is the biggest activation or a 1 for the netowrk or expected vals respectively
```
typo

kjhsdgf commented on Network Class/Network.cpp:
```diff
@@ -178,6 +176,60 @@ bool backProp(int index)
+
+	// biggest starts at 0 (not MIN) because any number < 0 might as well be 0 for these purposes
+	// numBiggest is number of times we encounter the biggest element in output layer
+	double biggest = 0;	int numBiggest = 0;
```
numBiggest can be a bool, and named to reflect that it represents a status.
Negative outputs may be a possibility, so set biggest to be equal to the value of the first element.

kjhsdgf commented on Network Class/Network.cpp:
```diff
@@ -178,6 +176,60 @@ bool backProp(int index)
+
+	// out of place random check to make sure it's being fed good data
+	if (expectedValues.size() != lastSize)
+		return 0;
```
Leave an error message.

kjhsdgf commented on Network Class/Network.cpp:
```diff
@@ -178,6 +176,60 @@ bool backProp(int index)
+		{
+			if (expectedI == -1) // if it's not been set before..
+				expectedI = i;
+			else				 // it it has it's bad data
```
typo

kjhsdgf commented on Network Class/Network.cpp:
```diff
@@ -178,6 +176,60 @@ bool backProp(int index)
+	 
+	// now we can be sure biggestI, expectedI >=0
+
+	if (numBiggest > 0) // if there's more than 1 biggest, ambiguous data. don't even care if the indices match
```
all the "bad data" cases return the same value. It may be worthwhile to distinguish the errors/ambiguities for troubleshooting.

kjhsdgf committed:
Updated to reflect my comments.

line 187: fixed typo (netowrk -> network)
changed numBiggest to a boolean variable, replaced mentions.
line 195: added error message to help in troubleshooting later on.
line 213: fixed typo (non-sequitor sentence)

kjhsdgf committed:
Merge branch 'master' into yonatanmg-compareOutput-method
